### PR TITLE
Audio file info

### PR DESCRIPTION
**Pull request by cspiel1**
This adds a command to debug_cmd that uses an ausrc (e.g. aufile) in order to count the samples and compute the length of the audio file in milli seconds.

The ausrc needs a new "blocking mode". As a first example we extend the aufile with a blocking mode.

Maybe later we also deliver our auogg (audio source for ogg files) or the gst module with a blocking mode.

### CODE REVIEW

**Comment by alfredh** [comment]
please note that libre/baresip is designed around an async event-driven polling loop.
one thread can handle hundreds of I/O events and timers.
for this to work properly, each function must execute fast.

I am sceptical about blocking the call for N seconds, this will basically break the whole
design pattern.

If you run this command on a long audio file (say, 30 minutes) the polling loop will
be blocked until the whole file is read. During this time, no other commands can be executed.

**Comment by cspiel1** [comment]
Good defence! I will try to find a better solution.

**Comment by cspiel1** [comment]
- removed the "blocking mode" in aufile
- now aufile has a mode with ptime is zero which processes as fast as possible, but does not block the alloc_handler
- In debug_cmd the cmd_fileinfo() informs with a ua_event() about the result. Thus module cons, ctrl_tcp or ctrl_dbus pass the result to the user asynchronously.

**Comment by alfredh** [comment]
I tested it locally on Debian 10, and it does not seem to work for me:

```
/fileinfo error.wav
.... dev '/usr/local/share/baresip/error.wav'
aufile: loading input file '/usr/local/share/baresip/error.wav'
aufile: /usr/local/share/baresip/error.wav: 8000 Hz, 1 channels, PCMU
aufile: audio ptime=0 sampc=320
aufile: read end of file
aufile: loaded 33436 bytes
debug_cmd: timeout, length > 1727557320 ms
```

File: modules/debug_cmd/debug_cmd.c [hunk 1]
```diff
@@ -18,7 +19,6 @@
  * Advanced debug commands
  */
 
-
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:21]: unrelated change ..

File: modules/debug_cmd/debug_cmd.c [hunk 2]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
 }
 
 
+struct ausrc_st;
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:163]: is this forward declaration needed ?

File: modules/debug_cmd/debug_cmd.c [hunk 3]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+{
+	struct fileinfo_st *st = arg;
+	size_t ms = 0;
+
+	if (st->prm.ch && st->prm.srate)
+		ms = st->sampc * 1000 / st->prm.ch / st->prm.srate;
+
+	if (st->finished) {
+		info("debug_cmd: length = %u ms\n", ms);
+		ua_event(NULL, UA_EVENT_MAX, NULL, "debug_cmd: length = %u ms",
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:194]: please dont use `UA_EVENT_MAX` for this purpose.

File: modules/debug_cmd/debug_cmd.c [hunk 4]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+static void fileinfo_timeout(void *arg)
+{
+	struct fileinfo_st *st = arg;
+	size_t ms = 0;
+
+	if (st->prm.ch && st->prm.srate)
+		ms = st->sampc * 1000 / st->prm.ch / st->prm.srate;
+
+	if (st->finished) {
+		info("debug_cmd: length = %u ms\n", ms);
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:193]: I think that the unit _seconds_ is more user-friendly, use e.g. `double seconds`.

length: 30500 ms
length: 30.5 seconds

File: modules/debug_cmd/debug_cmd.c [hunk 5]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+	else {
+		info("debug_cmd: timeout\n");
+		ua_event(NULL, UA_EVENT_MAX, NULL, "debug_cmd: timeout");
+	}
+
+	mem_deref(st);
+}
+
+
+static void fileinfo_readh(struct auframe *af, void *arg)
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:211]: the suffix for handlers should be `_handler`

same for below function.

File: modules/debug_cmd/debug_cmd.c [hunk 6]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+		return;
+
+	st->sampc += af->sampc;
+}
+
+
+static void fileinfo_errh(int err, const char *str, void *arg)
+{
+	struct fileinfo_st *st = arg;
+	(void) err;
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:225]: no need to cast err, it is used.

File: modules/debug_cmd/debug_cmd.c [hunk 7]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+ * config file_ausrc and returns the length in milli seconds. The file has
+ * to be located in the path specified by audio_path.
+ *
+ * Usage:
+ * /fileinfo audiofile
+ *
+ * @param pf Print handler is used to return length.
+ * @param arg Command argument contains the file name.
+ *
+ * @return The length of the fileplay file in milli seconds.
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:244]: is the return description correct ?

File: modules/debug_cmd/debug_cmd.c [hunk 8]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+		warning("debug_cmd: file_ausrc is not set\n");
+		return EINVAL;
+	}
+
+	len = str_len(conf_config()->audio.audio_path) + str_len(carg->prm)
+		+ 2;
+	path = mem_zalloc(len, NULL);
+	re_snprintf(path, len, "%s/%s",
+			conf_config()->audio.audio_path,
+			carg->prm);
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:272]: this block can be replaced with `re_sdprintf`

File: modules/debug_cmd/debug_cmd.c [hunk 9]
```diff
@@ -160,6 +160,143 @@ static int cmd_play_file(struct re_printf *pf, void *arg)
...
+	}
+
+	err = ausrc_alloc(&st->ausrc, baresip_ausrcl(),
+			NULL, aumod,
+			&st->prm, path,
+			fileinfo_readh, fileinfo_errh, st);
+	if (err)
+		warning("debug_cmd: %s - ausrc %s does not support zero ptime "
+				"or reading source %s failed. (%m)\n",
+				__func__, aumod, carg->prm, err);
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:288]: the error case should also goto out

File: modules/debug_cmd/debug_cmd.c [hunk 10]
```diff
@@ -226,6 +363,7 @@ static const struct cmd debugcmdv[] = {
 {"modules",     0,       0, "Module debug",           mod_debug           },
 {"netstat",    'n',      0, "Network debug",          cmd_net_debug       },
 {"play",        0, CMD_PRM, "Play audio file",        cmd_play_file       },
+{"fileinfo",    0, CMD_PRM, "Audio file info",        cmd_fileinfo        },
```
Comment (alfredh) [inline modules/debug_cmd/debug_cmd.c:366]: consider adding "au" prefix since it is audio related.

au - audio
vid - video

File: modules/gst/gst.c [hunk 11]
```diff
@@ -427,9 +435,21 @@ static int gst_alloc(struct ausrc_st **stp, const struct ausrc *as,
 	err = setup_uri(st, device);
 	if (err) goto out;
 
-	st->prm   = *prm;
+	st->ptime = prm->ptime;
+	if (!st->ptime)
+		st->ptime = 20;
+
+	if (!st->ptime)
+		st->ptime = 20;
```
Comment (alfredh) [inline modules/gst/gst.c:443]: duplicated code ?

**Comment by alfredh** [comment]
would it be possible to read out the duration field of the WAV header and use this ?

**Comment by cspiel1** [comment]
> would it be possible to read out the duration field of the WAV header and use this ?

Not with this approach. It works for general ausrc. E.g. module gst. The "zero ptime" extension in the ausrc modules let them process as fast as possible. And a timeout of 5 seconds stops if the file is to long or a stream was specified.

Thanks for the comments! I will update by adding commits.

Here is the output of my test with error.wav:
```/fileinfo error.wav
aufile: loading input file '/usr/local/share/baresip/error.wav'
aufile: /usr/local/share/baresip/error.wav: 8000 Hz, 1 channels, PCMU
aufile: audio ptime=0 sampc=320
aufile: read end of file
aufile: loaded 33436 bytes
aufile: end of file
debug_cmd: length = 2120 ms
```
I tried with gst and a very long mp3. Then I get this output:
```debug_cmd: timeout, length > 1272100 ms
```

File: modules/debug_cmd/debug_cmd.c (hunk 2 above)
Comment (cspiel1) [reply modules/debug_cmd/debug_cmd.c:163]: no.

File: modules/debug_cmd/debug_cmd.c (hunk 3 above)
Comment (cspiel1) [reply modules/debug_cmd/debug_cmd.c:194]: It would be nice in general to be able to send some kind of "general" events from any modules. 
- We could define for each purpose a constant e.g. UA_EVENT_FILEINFO, or
- we could define a constant in baresip.h called UA_EVENT_GENERIC.

File: modules/debug_cmd/debug_cmd.c (hunk 7 above)
Comment (cspiel1) [reply modules/debug_cmd/debug_cmd.c:244]: no it's not. Thanks!

File: modules/debug_cmd/debug_cmd.c (hunk 8 above)
Comment (cspiel1) [reply modules/debug_cmd/debug_cmd.c:272]: cool! ;-)

**Comment by cspiel1** [comment]
I suggest to squash the debug_cmd commits, the aufile commits and the gst commits separately. Ok?
